changes required for EIP-7742 (#574)
ethereum/execution-apis: src/engine/prague.md

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

> - [Structures](#structures)

Reviewer: gotta add execution payload v4 here I think

Author: ty, fixed

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Structures](#structures)
  - [ExecutionPayloadV4](#executionpayloadv4)
  - [PayloadAttributesV4](#payloadattributesv4)
- [Methods](#methods)
  - [engine_newPayloadV4](#engine_newpayloadv4)
    - [Request](#request)
    - [Response](#response)
    - [Specification](#specification)
  - [engine_forkchoiceUpdatedV4](#engine_forkchoiceupdatedv4)
    - [Request](#request-1)
    - [Response](#response-1)
    - [Specification](#specification-1)
  - [engine_getPayloadV4](#engine_getpayloadv4)
    - [Request](#request-2)
    - [Response](#response-2)
    - [Specification](#specification-2)
  - [Update the methods of previous forks](#update-the-methods-of-previous-forks)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Structures

> ### ExecutionPayloadV4

Reviewer 2: I don’t think we need a new ExecutionPayload definition, since we have `targetBlobsPerBlock` passed as a param to `newPayloadV4`

Author: that's the value coming from the CL, the EL needs a way to capture what is on-chain

the EL's job here is to verify the value in the EL header matches what the CL is sending

Reviewer 2: right, but the treatment of this value looks similar to `parentBeaconBlockRoot`, i don’t get why the `targetBlobsPerBlock` should be a part of the payload

Author: if we don't do it this way, we would need to have the CL validate this field in the `ExecutionPayload`

and I just took a look at how we handle the parent beacon block root, and kind of agree we should do this check there and can skip passing it via the `ExecutionPayloadV4` here

let's see if we decide to ship EIP-7742 in pectra and we can make these updates in that case

    - [Request](#request-1)
    - [Response](#response-1)
    - [Specification](#specification-1)

>   - [engine_getPayloadBodiesByHashV2](#engine_getpayloadbodiesbyhashv2)

Reviewer: I think you need to revert this deletion

Author: ? the TOC generator just moved them below

or are you talking about something else?

Reviewer: Ah sorry i see

- `withdrawalRequests`: `Array of WithdrawalRequestV1` - Array of withdrawal requests, each object is an `OBJECT` containing the fields of a `WithdrawalRequestV1` structure.
- `consolidationRequests`: `Array of ConsolidationRequestV1` - Array of consolidation requests, each object is an `OBJECT` containing the fields of a `ConsolidationRequestV1` structure.

### PayloadAttributesV4

> This structure has the syntax of `PayloadAttributesV3` and appends the fields: `targetBlobCount`, `maximumBlobCount`.

Reviewer 3: ```suggestion
This structure has the syntax of `PayloadAttributesV3` and appends the fields: `targetBlobsPerBlock`, `maximumBlobsPerBlock`.
```

- `timestamp`: `QUANTITY`, 64 Bits - value for the `timestamp` field of the new payload
- `prevRandao`: `DATA`, 32 Bytes - value for the `prevRandao` field of the new payload
- `suggestedFeeRecipient`: `DATA`, 20 Bytes - suggested value for the `feeRecipient` field of the new payload
- `withdrawals`: `Array of WithdrawalV1` - Array of withdrawals, each object is an `OBJECT` containing the fields of a `WithdrawalV1` structure.
- `parentBeaconBlockRoot`: `DATA`, 32 Bytes - Root of the parent beacon block.

> - `targetBlobCount`: `QUANTITY`, 64 Bits - Average number of blobs to include per payload.
> - `maximumBlobCount`: `QUANTITY`, 64 Bits - Maximum number of blobs allowed per payload.

Reviewer 3: ```suggestion
- `targetBlobsPerBlock`: `QUANTITY`, 64 Bits - Average number of blobs to include per payload.
- `maximumBlobsPerBlock`: `QUANTITY`, 64 Bits - Maximum number of blobs allowed per payload.
```

  1. `executionPayload`: [`ExecutionPayloadV4`](#ExecutionPayloadV4).
  2. `expectedBlobVersionedHashes`: `Array of DATA`, 32 Bytes - Array of expected blob versioned hashes to validate.
  3. `parentBeaconBlockRoot`: `DATA`, 32 Bytes - Root of the parent beacon block.

>   4. `targetBlobCount`: `QUANTITY`, 64 Bits - Average number of blobs to include per payload.

Reviewer 3: ```suggestion
  4. `targetBlobsPerBlock`: `QUANTITY`, 64 Bits - Average number of blobs to include per payload.
```

1. Client software **MUST** return `-38005: Unsupported fork` error if the `timestamp` of the built payload does not fall within the time frame of the Prague fork.

### engine_forkchoiceUpdatedV4

#### Request

* method: `engine_forkchoiceUpdatedV4`
* params:
  1. `forkchoiceState`: [`ForkchoiceStateV1`](./paris.md#ForkchoiceStateV1).
  2. `payloadAttributes`: `Object|null` - Instance of [`PayloadAttributesV4`](#payloadattributesv4) or `null`.
* timeout: 8s

#### Response

> Refer to the response for [`engine_forkchoiceUpdatedV3`](./shanghai.md#engine_forkchoiceupdatedv3).

Reviewer 3: ```suggestion
Refer to the response for [`engine_forkchoiceUpdatedV3`](./cancun.md#engine_forkchoiceupdatedv3).
```

> #### Specification

Reviewer 4: shouldn't this state that for `payloadAttributes`, `timestamp` must be within Prague hardfork?

> This method follows the same specification as [`engine_forkchoiceUpdatedV3`](./shanghai.md#engine_forkchoiceupdatedv3).

Reviewer 3: ```suggestion
This method follows the same specification as [`engine_forkchoiceUpdatedV3`](./cancun.md#engine_forkchoiceupdatedv3).
```
